src/aggregate/aggregate_exec.c

```diff
@@ -321,11 +331,45 @@ void sendChunk(AREQ *req, RedisModuleCtx *outctx, size_t limit) {
 }
 
 void AREQ_Execute(AREQ *req, RedisModuleCtx *outctx) {
-  sendChunk(req, outctx, -1);
-  if (IsProfile(req)) {
-    Profile_Print(outctx, req);
-  }
-  AREQ_Free(req);
+	if (IsProfile(req)) {
+    	RedisModule_ReplyWithArray(outctx, 2);
+    }
+	sendChunk(req, outctx, -1);
+	if (IsProfile(req)) {
+		Profile_Print(outctx, req);
+	}
+	AREQ_Free(req);
+}
+
+
+void AREQ_Execute_threadSafe(blockingClientReqCtx *BCRctx_) {
+	blockingClientReqCtx *BCRctx = (blockingClientReqCtx *)BCRctx_;
+	// lockspec
+	RedisSearchCtx_LockSpecRead(BCRctx->r->sctx);
+	AREQ_Execute(BCRctx->r, blockingClientReqCtx_getRedisctx(BCRctx));
+	// No need to unlock spec as AREQ_Execute calls ctx cleanup.
+
+	if(RedisModule_BlockedClientMeasureTimeEnd) {
+		// report block client end time
+		RedisModule_BlockedClientMeasureTimeEnd(BCRctx->bc);
+	}
+	blockingClientReqCtx_destroy(BCRctx);
+}
+
+blockingClientReqCtx *blockingClientReqCtx_New(AREQ *r, RedisModuleBlockedClient *bc) {
+  blockingClientReqCtx *ret = rm_malloc(sizeof(blockingClientReqCtx));
+  ret->r = r;
+  ret->bc = bc;
+  return ret;
+}
+
+RedisModuleCtx *blockingClientReqCtx_getRedisctx(const blockingClientReqCtx *BCRctx){
+  return BCRctx->r->sctx->redisCtx;
+}
+
+void blockingClientReqCtx_destroy(blockingClientReqCtx *BCRctx) {
+  RedisModule_UnblockClient(BCRctx->bc, NULL);
```

Comment:
Does this function require the GIL?

Reply:
https://redis.io/docs/reference/modules/modules-api-ref/#RedisModule_UnblockClient
Note 1: this function can be called from threads spawned by the module.